samza-core/src/main/scala/org/apache/samza/system/SystemConsumers.scala

```diff
@@ -237,6 +233,13 @@ class SystemConsumers (
     }
   }
 
+  def drain(): Unit = {
+    if (!isDraining) {
+      isDraining = true;
+      info("SystemConsumers is set to drain mode.")
+      writeDrainControlMessageToSspQueue()
```

Comment:
Given Kafka consumer will spawn up a thread, seems it's possible in a small chance to have more envelopes being added after the drain message. For example, the consumer thread is trying to pull the consumer, and we get the drain command at the same time. Is this situation being handled? It will be good to write that in the design doc.

Reply:
New envelopes will be added to SystemConsumers' in-memory queue from the registered consumers whenever refresh is called. When in drain mode, refresh will not add new messages from the consumers.